modules/gapi/test/serialization/serialization.cpp

```diff
@@ -0,0 +1,1292 @@
+            {
+                const auto idx = ade::util::index(out);
+                const auto rc = ade::util::value(out);
+
+                const auto& outDataNode = dataNodes[rc.shape][rc.id];
+                const auto& e = g.link(nh, outDataNode);
+                gm.metadata(e).set(Output{idx});
+            }
+        }
+    }
+    return nodes;
+}
+//} // anonymous namespace
+
+GSerialized serialize(const gimpl::GModel::ConstGraph& cg, const std::vector<ade::NodeHandle>& nodes)
+{
+    GSerialized s;
+    for (auto &nh : nodes)
+    {
+        switch (cg.metadata(nh).get<NodeType>().t)
+        {
+        case NodeType::OP:   putOp  (s, cg, nh); break;
+        case NodeType::DATA: putData(s, cg, nh); break;
+        default: util::throw_error(std::logic_error("Unknown NodeType"));
+        }
+    }
+    return s;
+}
+
+void deserialize(const serialization::GSerialized& s)
+{
+    printGSerialized(s);
+
+    // FIXME: reuse code from GModelBuilder/GModel!
+    // ObjectCounter?? (But seems we need existing mapping by shape+id)
+
+    ade::Graph g;
+
+    for (const auto& data : s.m_datas)
+    {
+        mkDataNode(g, data);
+    }
+
+    for (const auto& op : s.m_ops)
+    {
+        mkOpNode(g, op);
+    }
+
+    linkNodes(g);
+
+//  FIXME:
+//  Handle IslandModel!
+//    std::shared_ptr<ade::Graph> ig;
+//    GModel::Graph gm(g);
+//    gm.metadata().set(gimpl::IslandModel{std::move(ig)});
+
+    auto pass_ctx = ade::passes::PassContext{g};
+    ade::passes::TopologicalSort{}(pass_ctx);
+    gimpl::passes::dumpDotToFile(pass_ctx, "graph.dot");
+}
+
+/////////////////////////////////////////////////////////////////////////////////////
+//Graph dump operators
+template<typename _Tp> static inline
+SerializationStream&  operator << (SerializationStream& os, /*const*/ _Tp value)
+{
+    os << value;
+    return os;
+}
+
+SerializationStream& operator << (SerializationStream& os, Kernel k)
+{
+    CV_LOG_INFO(NULL, "k.name  " << k.name.c_str());
+    CV_LOG_INFO(NULL, "k.tag  " << k.tag.c_str());
+    os << k.name;
+    os << k.tag;
+    return os;
+}
+
+SerializationStream& operator << (SerializationStream& os, std::string str)
+{
+    uint str_size = (uint)str.size();
+    os << str_size;
+    size_t numAtoms = str_size % sizeof(uint) ==
+        0 ? (str_size / sizeof(uint)) : (str_size / sizeof(uint)) + 1;
+    uint* hton_buff = (uint*)malloc(numAtoms * sizeof(uint));
+    memcpy(hton_buff, str.c_str(), str_size);
+    for (uint a = 0; a < numAtoms; a++)
+    {
+        os << hton_buff[a];
+    }
+    return os;
+}
+
+SerializationStream& operator << (SerializationStream& os, std::vector<int> ints)
+{
+    os << (uint)ints.size();
+    for (auto & element : ints)
+    {
+        os << element;
+    }
+    return os;
+}
+
+SerializationStream& operator << (SerializationStream& os, std::vector<RcDesc> descs)
+{
+    os << (uint)descs.size();
+    for (auto & element : descs)
+    {
+        os << element;
+    }
+    return os;
+}
```

Comment:
Doesn't a template overload for vector work for all `T`?

Reply:
Also please take vectors as `const &`.

Reply:
Done. Not sure that i'm understand comment about template.